File: jiho/정복자.py
import heapq

N, M, t = map(int, input().split())
q = []
for i in range(M):
    a, b, c = map(int, input().split())
    heapq.heappush(q, [c, a, b])


def find_parent(parent, x):
    if parent[x] == x:
        return x

    parent[x] = find_parent(parent, parent[x])
    return parent[x]


def union(parent, x, y):
    X = find_parent(parent, x)
    Y = find_parent(parent, y)
    if X == Y:
        return False
    parent[X] = Y
    return True


def solution():
    global N, M, t, q
    parent = [i for i in range(N + 1)]

    result, cnt = 0, 0
    while cnt < N - 1:
        c, x, y = heapq.heappop(q)
        if union(parent, x, y):
            result += c + t * cnt
            cnt += 1

    return result


print(solution())

File: cjeongmin/정복자.js
const input = require('fs').readFileSync(0, 'utf-8').toString().trim().split('\n');

const [N, M, T] = input[0].split(' ').map(Number);

const edges = [];
for (let i = 1; i <= M; i++) {
  const [a, b, c] = input[i].split(' ').map(Number);
  edges.push([c, a, b]);
}

edges.sort((a, b) => a[0] - b[0]);
const parent = Array.from({ length: N + 1 }, (_, i) => i);

const find = (x) => {
  if (parent[x] === x) return x;
  return (parent[x] = find(parent[x]));
};

const union = (a, b) => {
  const rootA = find(a);
  const rootB = find(b);

  if (rootA !== rootB) {
    if (rootA < rootB) {
      parent[rootB] = rootA;
    } else {
      parent[rootA] = rootB;
    }
    return true;
  }
  return false;
};

let result = 0;
let cnt = 0;

for (const [cost, a, b] of edges) {
  if (union(a, b)) {
    result += cost + cnt * T;
    cnt++;

    if (cnt === N - 1) {
      break;
    }
  }
}

console.log(result);
